ingestion/src/main.rs

```diff
@@ -66,25 +67,36 @@ async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
         }
     });
 
-    // Spawn kvkafka reader
-    #[cfg(feature = "kafka_prod")]
-    {
-        let kafka_group = args[1].to_string();
-        println!("Spawing kvkafka reader...");
-        tokio::spawn(lard_ingestion::kvkafka::read_and_insert(
-            db_pool.clone(),
-            kafka_group,
-        ));
-    }
+    // set up cancellation token and signal catcher for graceful shutdown
+    let cancel_token = CancellationToken::new();
+    tokio::spawn(util::signal_catcher(cancel_token.clone()));
 
     // Set up and run our server + database
     println!("Ingestion server started!");
-    lard_ingestion::run(
-        db_pool,
+    let ingestor = tokio::spawn(lard_ingestion::run(
+        db_pool.clone(),
         PARAMCONV,
         permit_tables,
         rove_connector,
         qc_pipelines,
-    )
-    .await
+        cancel_token.clone(),
+    ));
+
+    #[cfg(feature = "kafka_prod")]
+    // Spawn kvkafka reader
+    {
+        let kafka_group = args[1].to_string();
+        println!("Spawning kvkafka reader...");
+        let kvkafka_reader = tokio::spawn(lard_ingestion::kvkafka::read_and_insert(
+            db_pool,
+            kafka_group,
+            cancel_token,
+        ));
+
+        let (ingestor_res, kvkafka_reader_res) = tokio::join!(ingestor, kvkafka_reader);
+        ingestor_res.unwrap().map(|_| kvkafka_reader_res.unwrap())
```

Comment:
Not what I meant. Here you panic on error instead of returning it

Reply:
I thought it didn't matter since the program was exiting at this point anyway, and I assumed the details would end up in some log (stderr?). I actually tried the .and_then() idiom first, but ran into some problems. I'll take another look.

Reply:
(and I guess I shouldn't just assume things, but actually test it ;) )

Reply:
Isn't it panicking on panic, since `join` returns `Result<T, JoinError>`? The actual error should be inside `T`, no? But I guess we can still use question marks instead, and then you wouldn't need to use the map function.

Reply:
Done!